### PR TITLE
missions: move Old Guard rep mission start functions to JSON

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Infrastructure "missions: move Old Guard rep mission start functions to JSON"```
#### Purpose of change
Make it possible for JSON only contributors to update existing overmap tiles when starting NPC missions.
In the process, develop and test an generic infrastructure for updating existing overmap tiles through JSON.

#### Describe the solution
Create a new update_mapgen object in mission JSON definition.  update_mapgen contains standard mapgen JSON objects that are applied to an existing overmap tile.

#### Additional context
Part of #26794
I plan to move more mission starts into JSON later but I want to get this out and reviewed.

### CODE REVIEW

File: src/mapgen.cpp [hunk 1]
```diff
@@ -2155,8 +2190,11 @@ void jmapgen_objects::check( const std::string &oter_name ) const
  * (set|line|square)_(ter|furn|trap|radiation); simple (x, y, int) or (x1,y1,x2,y2, int) functions
  * TODO: optimize, though gcc -O2 optimizes enough that splitting the switch has no effect
  */
-bool jmapgen_setmap::apply( const mapgendata &dat, int offset_x, int offset_y ) const
+bool jmapgen_setmap::apply( const mapgendata &dat, int offset_x, int offset_y, mission *miss ) const
 {
+    if( miss )
+        ;
```
Comment (codemime) [inline src/mapgen.cpp]: This if weird. You could simply omit `miss` in the signature. Or is it a leftover of something?

Comment (mlangsdorf) [reply src/mapgen.cpp]: leftovers, thanks for catching it

Comment (kevingranade) [reply src/mapgen.cpp]: There are quite a few of these, and they trigger warnings.

**Comment by ralreegorganon** [comment]
Tested this out, works well for the missions in question, but noticed some things:

1. The offset documented [here](https://github.com/CleverRaven/Cataclysm-DDA/pull/29294/files#diff-3e6abf55e2fd2347508f11dfc502170eR709) is a single integer but should probably be a point, and is implied as being a row offset but is actually a column offset. It's being used [here](https://github.com/CleverRaven/Cataclysm-DDA/pull/29294/files#diff-f4dada645aa2c6343e16a20b047c07c5R8370) and [here](https://github.com/CleverRaven/Cataclysm-DDA/pull/29294/files#diff-f4dada645aa2c6343e16a20b047c07c5R8372) but just being passed into offset_x and offset_y is kept at zero. Making it a point will allow offsets on both axis.

2. This doesn't handle updating rotated locations--it works for the ones in question because they all have rotate false, but an update against a rotated location will end up placing things in the wrong location. The normal mapgen process basically does the mapgen and then rotates the result to the correct orientation, but I don't think we want to un-rotate and re-rotate here and so we'll need a different approach. That seems like a big enough task (and most of the existing mission starts we'd jsonize are for non-rotating locations) that we should just include the caveat right now that this only works for locations that can't rotate, and deal with rotation later.

3. This isn't even a bug with these changes, but rather an assumption from the current implementation that I violated when jsonizing the mission start you get from the computer in the evac shelter that reveals the center (`mission_start::reveal_refugee_center`): [this code](https://github.com/CleverRaven/Cataclysm-DDA/pull/29294/files#diff-5e5bc5383145a04ba3ee985cd8f84877L2000) is assuming that you have an NPC, but in the case of the mission from the computer, you don't have one.

I think 1 is worth fixing now, but 2 seems like a follow up task and 3 can easily be dealt with later as well.

**Comment by mlangsdorf** [comment]
Fixed item 1.
Added some docs about item 2.
Fixed item 3.

Thanks for the review and testing.

**Comment by ralreegorganon** [comment]
Did some more testing of this following the updates.

First, maybe it's just Visual Studio (I don't have my other environments on hand to test right now), but I get a crash on application startup with the fix you added for the default NPC:

image

Second, for the offset, the docs and code are out of sync as far the name of the parameters. The [docs say "offset_x"](https://github.com/CleverRaven/Cataclysm-DDA/pull/29294/files#diff-3e6abf55e2fd2347508f11dfc502170eR708) while the [code reads "x_offset"](https://github.com/CleverRaven/Cataclysm-DDA/pull/29294/files#diff-f4dada645aa2c6343e16a20b047c07c5R8379). 

Sorry if I wasn't clear in the previous feedback regarding using a point with the "offset" variable--I was thinking you'd keep the name `"offset"` and update the type, e.g. `"offset": [12, 12]` and reading straight into the point with `jo.read("offset", offset);`. At least in my mind, it's more conventional because the offset is really just a new origin point and so might as well just be represented as one, but maybe that's just me.

However, I'm actually wondering if we shouldn't just strike the offset entirely for this first pass. It's not used by the missions you jsonized here, and after doing some testing with it, I'm not sure it's a great experience without also allowing/requiring specification of `mapgensize` like the nested mapgen does. My rationale is as follows:
* The intent of specifying an offset here is to allow a sparser definition rather than requiring the entire 24x24 mapgen entry be specified
* If you're just placing things using the `place_X` functions, you don't have to specify every location, and so all the offset does is shift the x and y coordinates
* If you are placing things using the `"rows"` json array, the offset only solves half the problem--if you want to define a 1x1 square at 1, 1, you can define an offset of [1, 1], but you actually have to define 23 columns and rows rather than just the 1 because without changing the `mapgensize`, the rows plus the offset has to equal 24
* Placing a nested mapgen will cover the sparse placement. It's maybe a little unfortunate that it's not entirely inline in the update_mapgen, but it works great and accomplishes the end goal. For example, here's an existing nested mapgen, and how you could place it in an update:
```json
"update_mapgen": {
  "place_nested": [
    { "chunks": [ "roof_4x4_holdout" ], "x": 12, "y": 12 }
  ]
}
...
{
  "type": "mapgen",
  "method": "json",
  "nested_mapgen_id": "roof_4x4_holdout",
  "object": {
    "mapgensize": [ 4, 4 ],
    "rows": [ 
      "xxxx", 
      "xct_", 
      "x___", 
      "____" 
    ],
    "terrain": { "_": "t_flat_roof" },
    "furniture": { "c": "f_camp_chair", "x": "f_sandbag_wall", "t": "f_table" },
    "place_loot": [
      { "group": "roof_trash", "x": [ 0, 3 ], "y": 3, "chance": 30, "repeat": [ 1, 2 ] },
      { "group": "alcohol_bottled_canned", "x": [ 1, 3 ], "y": 3, "chance": 80 },
      { "group": "stash_food", "x": 2, "y": 1, "chance": 50, "repeat": [ 2, 3 ] },
      { "group": "roof_holdout", "x": [ 1, 3 ], "y": [ 1, 3 ], "chance": 70, "repeat": [ 2, 4 ] }
    ]
  }
}
```

**Comment by mlangsdorf** [comment]
Nested mapgen is a much better solution.  I do want to use minimal maps, but if nested mapgen chunks work with update_mapgen, that's a better solution than reinventing the wheel.

Fixed the crash with the reference NPC.

**Comment by ralreegorganon** [comment]
Gave it a spin with the latest updates, doing my test case of jsonizing the start for `MISSION_REACH_REFUGEE_CENTER`. No more crashes, looks good. :+1:

**Pull request merged**